src/main/scala/com/evolutiongaming/cluster/ExtendedShardAllocationStrategy.scala

```diff
@@ -0,0 +1,76 @@
+package com.evolutiongaming.cluster
+
+import java.time.Instant
+
+import akka.actor.{ActorRef, ActorSystem, Address}
+import akka.cluster.sharding.ShardCoordinator.ShardAllocationStrategy
+import akka.cluster.sharding.ShardRegion
+
+import scala.collection.immutable
+import scala.concurrent.{ExecutionContext, Future}
+import scala.concurrent.duration._
+
+abstract class ExtendedShardAllocationStrategy(
+  system: ActorSystem,
+  implicit val ec: ExecutionContext,
+  maxSimultaneousRebalance: Int,
+  deallocationTimeout: FiniteDuration) extends ShardAllocationStrategy {
+
+  val addressHelper = AddressHelperExtension(system)
+  import addressHelper._
+
+  class Node(val address: Address, added: Instant = Instant.now()) {
+    def expired: Boolean = Instant.now() isAfter (added plusMillis deallocationTimeout.toMillis)
+    override def equals(obj: Any): Boolean = obj match {
+      case node: Node => address equals node.address
+      case _          => false
+    }
+    override def hashCode(): Int = address.hashCode
+  }
+
+  @volatile
+  private var nodesToForcedDeallocation: Set[Node] = Set.empty
+
+  def deallocateShardsFromNode(regionGlobalAddress: Address): Unit =
+    nodesToForcedDeallocation += new Node(regionGlobalAddress)
```

Comment:
what cluster keep running without downtime, however we are continually restarting nodes ?